stl/inc/memory

```diff
@@ -1522,6 +1558,228 @@ private:
+
+    auto _Get_ptr() {
+        return reinterpret_cast<remove_extent_t<_Ty>*>(
+            reinterpret_cast<char*>(this) + _Div_ceil(sizeof(*this), alignof(_Ty)) * alignof(_Ty));
+    }
+
+private:
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        // nothing to do, _Ty is trivially destructible
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        using _Storage = aligned_storage_t<(_STD max)(alignof(_Ty), alignof(_Ref_count_unbounded_array<_Ty>)),
+            (_STD max)(alignof(_Ty), alignof(_Ref_count_unbounded_array<_Ty>))>;
+        this->~_Ref_count_unbounded_array();
+        delete[] reinterpret_cast<_Storage*>(this);
+    }
+};
+
+template <class _Ty>
+class _Ref_count_unbounded_array<_Ty, false> : public _Ref_count_base {
+    // handle reference counting for unbounded array with non-trivial destruction in control block, no allocator
+public:
+    explicit _Ref_count_unbounded_array(const size_t _Count) : _Ref_count_base(), _Size(_Count) {
+        _Uninitialized_value_construct_multidimensional_n(_Get_ptr(), _Size);
+    }
+
+    explicit _Ref_count_unbounded_array(const size_t _Count, const remove_extent_t<_Ty>& _Val)
+        : _Ref_count_base(), _Size(_Count) {
+        _Uninitialized_fill_multidimensional_n(_Get_ptr(), _Size, _Val);
+    }
+
+    ~_Ref_count_unbounded_array() = default; // nothing to do, _Storage was already destroyed in _Destroy
+
+    auto _Get_ptr() {
+        return reinterpret_cast<remove_extent_t<_Ty>*>(
+            reinterpret_cast<char*>(this) + _Div_ceil(sizeof(*this), alignof(_Ty)) * alignof(_Ty));
+    }
+
+private:
+    size_t _Size;
+
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        _Reverse_destroy_multidimensional_n(_Get_ptr(), _Size);
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        using _Storage = aligned_storage_t<(_STD max)(alignof(_Ty), alignof(_Ref_count_unbounded_array<_Ty>)),
+            (_STD max)(alignof(_Ty), alignof(_Ref_count_unbounded_array<_Ty>))>;
+        this->~_Ref_count_unbounded_array();
+        delete[] reinterpret_cast<_Storage*>(this);
+    }
+};
+
+// CLASS TEMPLATE _Ref_count_bounded_array
+template <class _Ty>
+class _Ref_count_bounded_array : public _Ref_count_base {
+    // handle reference counting for bounded array in control block, no allocator
+public:
+    _Ref_count_bounded_array() : _Ref_count_base(), _Storage() {}
+
+    explicit _Ref_count_bounded_array(const remove_extent_t<_Ty>& _Val) : _Ref_count_base() {
+        _Uninitialized_fill_multidimensional_n(_Storage._Value, extent_v<_Ty>, _Val);
+    }
+
+    ~_Ref_count_bounded_array() {
+        // nothing to do, _Storage was already destroyed in _Destroy
+    }
+
+    union {
+        _Wrap<_Ty> _Storage;
+    };
+
+private:
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        _Destroy_in_place(_Storage);

```

Comment:
The destructor comment and `_Destroy` mention `_Storage`, but the constructor initializes `_Storage._Value` and that's what `_Ref_count_obj2` destroys. I think this should be consistent. (The `_Wrap` struct exists only to handle the non-Standard `T^` extension.)

Reply:
Aha, you wrote that intentionally:

```
xmemory(269,9): error: object expression of non-scalar type 'std::basic_string<char, std::char_traits<char>, std::allocator<char> > [100]' cannot be used in a pseudo-destructor expression
    _Obj.~_Ty();
    ~~~~^
memory(1766,9): note: in instantiation of function template specialization 'std::_Destroy_in_place<std::basic_string<char, std::char_traits<char>, std::allocator<char> > [100]>' requested here
        _Destroy_in_place(_Storage._Value);
        ^
```

I'll revert my change, but add a comment. (As a general rule, we comment anything that has to diverge from a pattern followed elsewhere.)

Reply:
Arrays are quite a mess currently. I really hope P1997R1 will be approved for C++23 (there probably isn't a chance for C++20) to fix them (one of the things this paper proposes is to give arrays a pseudo-destructor so this array wouldn't have to be wrapped to have one)

Reply:
I'd suggest using `destroy_at` instead of `_Destroy_in_place`, but we haven't implemented the change to teach it about array types (IIRC from P0896).